File: 12-proxy/12-proxy.php
<?php

interface Server
{
    public function access();
}

class SomeServer implements Server
{
    public function access()
    {
        return "access granted";
    }
}

class ServerProxy implements Server
{
    public function __construct(
        private SomeServer $someServer,
        private bool $isAdmin
    ) {
    }

    public function access()
    {
        if (!$this->isAdmin) {
            return "access denied";
        }

        return $this->someServer->access();
    }
}

$some = new SomeServer();
$admin = new ServerProxy($some, true);
echo $admin->access();

echo "\n";

$nonAdmin = new ServerProxy($some, false);
echo $nonAdmin->access();

File: 12-proxy/12-proxy.py
class Server:
    def access(self) -> str:
        return "access granted"

class ServerProxy:
    def __init__(self, server: Server, is_admin: bool):
        self.server = server
        self.is_admin = is_admin

    def access(self) -> str:
        if self.is_admin == False:
            return "access denied"
        
        return self.server.access()

s = Server()
admin = ServerProxy(s, True)
print("admin:", admin.access())

guest = ServerProxy(s, False)
print("guest:", guest.access())
